tools/codegen/gen_vmap_plumbing2.py

```diff
@@ -0,0 +1,213 @@
+    for arg in flat_args:
+        if not arg.type.is_tensor_like():
+            continue
+        conditions.append(f'!isBatchedAtLevel({arg.name}, cur_level)')
+    aten_op = get_aten_op_call(schema)
+    arg_names = [a.name for a in flat_args]
+
+    return f"""\
+if ({' && '.join(conditions)}) {{
+  return {aten_op}({', '.join(arg_names)});
+}}"""
+
+
+def gen_returns(returns: List[Return]) -> str:
+    idx = 0
+    wrapped_returns = []
+    for ret in returns:
+        if is_tensor(ret.type):
+            wrapped_returns.append(f'makeBatched(std::get<{idx}>(results), std::get<{idx + 1}>(results), cur_level)')
+            idx += 2
+        elif is_vector_tensor(ret.type):
+            wrapped_returns.append(
+                f'makeBatchedVector(std::get<{idx}>(results), std::get<{idx + 1}>(results), cur_level)'
+            )
+            idx += 2
+        else:
+            wrapped_returns.append(f'std::get<{idx}>(results)')
+            idx += 1
+    if len(wrapped_returns) == 1:
+        wrapped_returns = f'return {wrapped_returns[0]};'
+    else:
+        wrapped_returns = f'return std::make_tuple({", ".join(wrapped_returns)});'
+    return wrapped_returns
+
+
+def accepts_at_least_one_tensor_input(schema):
+    for arg in schema.arguments.flat_all:
+        if arg.type.is_tensor_like():
+            return True
+    return False
+
+
+def gen_vmap_inplace_plumbing(native_function):
+    schema = native_function.func
+    sig = DispatcherSignature.from_schema(schema)
+    returns = schema.returns
+
+    assert schema.kind() == SchemaKind.inplace
+
+    # Only support cases where all returns are Tensors or vector<Tensor>
+    if len(returns) == 0:
+        return None
+    if not all(is_tensor(ret.type) or is_vector_tensor(ret.type) for ret in returns):
+        return None
+    if not accepts_at_least_one_tensor_input(schema):
+        return None
+
+    unwraps, unwrapped_arg_list = gen_unwraps(schema.arguments.flat_all)
+    bdims_all_none_case = gen_case_where_all_bdims_are_none(schema.arguments.flat_all, schema)
+
+    return f"""\
+template <typename batch_rule_t, batch_rule_t batch_rule>
+{sig.decl(name=schema.name.unambiguous_name() + '_generated_plumbing')} {{
+  c10::impl::ExcludeDispatchKeyGuard guard(kBatchedKey);
+  auto maybe_layer = maybeCurrentDynamicLayer();
+  TORCH_INTERNAL_ASSERT(maybe_layer.has_value());
+  int64_t cur_level = maybe_layer->layerId();
+{textwrap.indent(bdims_all_none_case, "  ")}
+{textwrap.indent(unwraps, "  ")}
+  batch_rule({', '.join(unwrapped_arg_list)});
+  return {schema.arguments.flat_all[0].name};
+}}"""
+
+
+def gen_vmap_plumbing(native_function: NativeFunction) -> str:
+    schema = native_function.func
+    sig = DispatcherSignature.from_schema(schema)
+    returns = schema.returns
+
+    # Only support cases where all returns are Tensors or vector<Tensor>
+    if len(returns) == 0:
+        return None
+    if not all(is_tensor(ret.type) or is_vector_tensor(ret.type) for ret in returns):
```

Comment:
I think `all(ret.type.is_tensor_like() for ret in returns)` will do what you want here

Reply:
yep it should - you can see e.g. [here](https://github.com/pytorch/pytorch/blob/e6afa4f77192830c4037448fdba3cf42a5c7bb28/tools/codegen/model.py#L1211) - we have a bunch of different `Type` subclasses, and both `OptionalType` and `ListType` implement `is_tensor_like()` by asking if their underlying type is tensor-like. So `ListType(OptionalType(BaseType(BaseTy.Tensor)), ...).is_tensor_like()` should just do the right thing.